### DISCUSSION TITLE
TypeAdapters are not reused leading to significant memory usage.

### DISCUSSION THREAD

**Discussion by atti92** [fastapi/fastapi#12934] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

**fastapi/_compat.py:109 in __post_init__**
```python

# This creates a new TypeAdapter for each endpoint's ModelField, which leads to excessive memory usage.
# Pydantic documentation #discourages the recreation of new typeadapters because they are memory heavy,
# and also computational heavy on creation.

#I suggest caching each model type, so it's not creating new objects.


#Example hack I tried to check if this makes a difference.

    @dataclass
    class ModelField:
        field_info: FieldInfo
        name: str
        mode: Literal["validation", "serialization"] = "validation"
        _typeadapters = {}

        def __post_init__(self) -> None:
            key = f"{self.field_info.annotation}, {self.field_info}"
            if key not in ModelField._typeadapters:
                ModelField._typeadapters[key] = TypeAdapter(
                    Annotated[self.field_info.annotation, self.field_info]
                )
            self._type_adapter: TypeAdapter[Any] = ModelField._typeadapters[key]
```

### Description

We've noticed in our company that adding more and more endpoints to our server started to increase the memory. 

We used to have below 200MB usage, now even starting up the service puts it at 800MB. Running some memray I've noticed how much TypeAdapters were allocated, and began to experiment. I realized fastapi is not reusing the typeadapters it creates, thos leading to wasted memory. 

Based on pydantic docs, I think it's a bad pattern to create a new object for each endpoint, especially in our case, when having mostly repetitive models which are quite big and complex.

Doing a simple caching inside the ModelField class I could reduce out initial 800MB usage to just 200MB

### Operating System

Linux

### FastAPI Version

0.115.4

### Pydantic Version

2.9.2

### Python Version

3.12.5

**Comment by dolfinus** [comment]
Does #12184 solve this issue?

**Comment by yakubka** [comment]
Creating a `TypeAdapter` inside a function means a new one is built on every call. `TypeAdapter` construction is not free: Pydantic v2 analyzes the type, builds a core schema, and compiles a validator on instantiation. For complex types this can take several milliseconds and allocate significant memory.

The fix is to create `TypeAdapter` instances at module level so they are built once.

```python
from pydantic import TypeAdapter
from typing import list

# built once at import time
ItemListAdapter = TypeAdapter(list[Item])

@app.get("/items")
async def get_items() -> list[Item]:
    raw = fetch_raw_data()
    return ItemListAdapter.validate_python(raw)
```

For types that depend on runtime values (generic parameters resolved at runtime), you need to cache them manually using a dict keyed by the type.

```python
from functools import lru_cache

@lru_cache(maxsize=128)
def get_adapter(item_type: type) -> TypeAdapter:
    return TypeAdapter(item_type)
```

`lru_cache` works here because `TypeAdapter` is not hashable by default but the type argument is. This avoids rebuilding the adapter on repeated calls with the same type.
